File: my-reader-mobile/modules/readium/android/src/main/java/com/myreader/readium/Streamer/PublicationStore.kt
package com.myreader.readium.Streamer

import com.myreader.readium.Search.SearchSessionStore
import org.readium.r2.shared.publication.Publication

/**
 * id+handle registry for stateful `Publication` objects (REP-003).
 *
 * JS holds a `publicationId` (the file URL) received from `onPublicationReady`
 * and passes it to module functions (`getPublicationSnapshot`, `getContent`,
 * `search`). Native resolves the id back to the live `Publication` here.
 */
object PublicationStore {
  private val lock = Any()
  private val store = mutableMapOf<String, Publication>()

  fun set(id: String, publication: Publication) {
    synchronized(lock) {
      val previous = store[id]
      if (previous != null && previous !== publication) {
        SearchSessionStore.cancelPublication(id)
      }
      store[id] = publication
    }
  }

  fun get(id: String): Publication? = synchronized(lock) { store[id] }

  fun remove(id: String, ifSameAs: Publication? = null) {
    synchronized(lock) {
      val current = store[id]
      if (current != null && (ifSameAs == null || current === ifSameAs)) {
        store.remove(id)
        SearchSessionStore.cancelPublication(id)
      }
    }
  }
}

File: my-reader-mobile/modules/readium/ios/Streamer/PublicationStore.swift
import Foundation
import ReadiumShared

/// id+handle registry for opened `Publication` objects (REP-003/004).
///
/// The id is the `bookId` (= file URL) emitted in `onPublicationReady`. JS
/// holds that id and calls module-level `getPublicationSnapshot`/`getContent`/
/// `search` through it. This avoids Expo SharedObject while keeping the object
/// graph explicit and reachable for TTS content iteration and services.
final class PublicationStore {
  static let shared = PublicationStore()

  private var publications: [String: Publication] = [:]
  private let queue = DispatchQueue(label: "my-reader.readium.publication-store")

  private init() {}

  func set(_ id: String, _ publication: Publication) {
    queue.sync {
      if let previous = publications[id], previous !== publication {
        SearchSessionStore.shared.cancel(publicationId: id)
      }
      publications[id] = publication
    }
  }

  func get(_ id: String) -> Publication? {
    queue.sync { publications[id] }
  }

  func remove(_ id: String, ifSameAs publication: Publication? = nil) {
    queue.sync {
      guard publication == nil || publications[id] === publication else {
        return
      }
      if publications.removeValue(forKey: id) != nil {
        SearchSessionStore.shared.cancel(publicationId: id)
      }
    }
  }
}
